### PR TITLE
ecds: hook UDP session filters with config provider

### PR DESCRIPTION
**Pull request by ohadvano**
Additional Description: Adding a new ``ConfigProviderManager`` for UDP session filters with tests. Consuming this provider for static filters. This PR does not add dynamic filters yet, will be done in next PR.
Risk Level: low
Testing: unit tests
Docs Changes: none
Release Notes: none
Platform Specific Features: none

### CODE REVIEW

**Comment by ohadvano** [comment]
@RyanTheOptimist could you please take this one, as you review the previous?

**Comment by adisuissa** [comment]
Assigning Ryan, but please let me know if there's someone else that you wish to review this.
/assign @RyanTheOptimist

**Comment by RyanTheOptimist** [comment]
/assign @htuch 
Harvey, you're a lot more familiar with xDS than I am. Can you review this?

**Comment by adisuissa** [comment]
> /assign @htuch Harvey, you're a lot more familiar with xDS than I am. Can you review this?

I should probably review it, and pass it to Kuat later.
Unassigning Harvey for now.

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

/assign-from @envoyproxy/senior-maintainers

**Comment by adisuissa** [comment]
Also assigning Kuat who has lots of experience.
/assign @kyessenov

File: source/extensions/filters/udp/udp_proxy/config.h [hunk 1]
```diff
@@ -151,8 +158,12 @@ class UdpProxyFilterConfigImpl : public UdpProxyFilterConfig,
 
   // UdpSessionFilterChainFactory
   void createFilterChain(Network::UdpSessionFilterChainFactoryCallbacks& callbacks) const override {
-    for (const Network::UdpSessionFilterFactoryCb& factory : filter_factories_) {
-      factory(callbacks);
+    for (const auto& filter_config_provider : filter_factories_) {
+      auto config = filter_config_provider->config();
+      if (config.has_value()) {
```
Comment (kyessenov) [inline source/extensions/filters/udp/udp_proxy/config.h:163]: Should there be some warning/error for a missing config here?

**Review by kyessenov — CHANGES_REQUESTED** [review]

File: source/extensions/filters/udp/udp_proxy/config.h (hunk 1 above)
Comment (ohadvano) [reply source/extensions/filters/udp/udp_proxy/config.h:163]: agree, added a log

Comment (kyessenov) [reply source/extensions/filters/udp/udp_proxy/config.h:163]: Thanks. Can you update the docs in the UDP listener, saying that the missing config in ECDS results in a no-op warning? For other filters, it is protocol-specific, e.g. 503 for HTTP.

Comment (ohadvano) [reply source/extensions/filters/udp/udp_proxy/config.h:163]: ECDS isn't really getting supported in this PR - the API remains the same, requiring consumers to configure the filter statically. In next PR I'll add dynamic consumption that uses discovery, and I'll update docs accordingly.
Also note that this kind of change applies to UDP **session** filters and not regular UDP listener filters

**Review by kyessenov — APPROVED** [review]

**Comment by adisuissa** [comment]
Seems that coverage needs to be improved.
See https://storage.googleapis.com/envoy-pr/02302c0/coverage/source/extensions/filters/udp/udp_proxy/index.html

Also adding @htuch for senior-maintainer stamp instead of @wbpcode who is OOO.
/assign @htuch

**Comment by ohadvano** [comment]
@kyessenov I removed the log line I added for two reasons:
1. Coverage is reduced below minimum, and I'm not sure if this scenario is reproducible in test (see 2nd reason).
2. Not sure if we would ever get there. For static filters - the filter factory validates the protobuf message, so if the config is missing, an exception will be thrown at bootstrap. On dynamic updates (not functional yet), xDS will also throw an exception if config update is invalid (factory will also try to instantiate the filter and fail)

**Comment by ohadvano** [comment]
@adisuissa did you mean to assign @htuch?
No senior maintainer is currently assigned

**Comment by adisuissa** [comment]
> @adisuissa did you mean to assign @htuch? No senior maintainer is currently assigned

Nope.. race between repokitteh and github's UI. sorry.

**Comment by kyessenov** [comment]
> @kyessenov I removed the log line I added for two reasons:
> 
> 1. Coverage is reduced below minimum, and I'm not sure if this scenario is reproducible in test (see 2nd reason).
> 2. Not sure if we would ever get there. For static filters - the filter factory validates the protobuf message, so if the config is missing, an exception will be thrown at bootstrap. On dynamic updates (not functional yet), xDS will also throw an exception if config update is invalid (factory will also try to instantiate the filter and fail)

This path used when ECDS fails to push config and the parent LDS was not configured to warm on ECDS. I think we have to handle this case, but I'm OK waiting till the final integration.

**Comment by ohadvano** [comment]
I see, makes sense. I'll add it back once in the dynamic configs PR with test that would be able to cover it using ECDS

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**